DRILL-7711: Add data path, parameter filter pushdown to HTTP plugin

paul-rogers opened the pull request:
# [DRILL-7711](https://issues.apache.org/jira/browse/DRILL-7711): Add data path, parameter filter pushdown to HTTP plugin

## Description

Adds an option to specify the path to data so the plugin will ignore REST message "overhead" except the actual data.

Allows specifying HTTP URL parameters as filter push-downs from SQL.

Includes improvements to the JSON structure parser to handle a few more conditions. Also includes a generic filter push-down framework for the case where filters are translated to expressions as here. (The framework is not applicable to cases where expressions are interpreted, as in partition pruning.)

## Documentation

Please see the [README.md](https://github.com/apache/drill/blob/0c01fb619d869c9cc0fd8ee094bc4ddd9042b552/contrib/storage-http/README.md) file for full details.

## Testing

Added unit tests for new functionality. Reran all existing tests.

arina-ielchiieva commented on contrib/storage-http/README.md:
````diff
@@ -29,41 +31,193 @@ The required options are:
+`myConn`. First, can include them directly in your URL if the parameters a fixed for a given
+service:
+
+```json
````
I think this notation does really work with markdown, please see doc display:
https://github.com/paul-rogers/drill/blob/a4226a06361bc136427797f542c401e05d608c5f/contrib/storage-http/README.md

paul-rogers replied:
Actually, it does work. The link I included was to an earlier version that had formatting errors. (I just learned that the link does not update when I push an update. Changed it by hand.) [New version](https://github.com/apache/drill/blob/0c01fb619d869c9cc0fd8ee094bc4ddd9042b552/contrib/storage-http/README.md).

To be honest, I didn't know about this feature until I saw what Charles had done. But, [it does work](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#code) and [many renderers](https://highlightjs.org/static/demo/) are supported.

arina-ielchiieva replied:
Great, new version looks much better.

arina-ielchiieva commented on contrib/storage-http/README.md:
````diff
@@ -29,41 +31,193 @@ The required options are:
+
+Query your table like this:
+
+```sql
````
Same here and for all other similar cases.

arina-ielchiieva commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/filter/ConstantHolder.java:
```diff
@@ -0,0 +1,381 @@
+    case VARDECIMAL:
+      return ((BigDecimal) value).compareTo((BigDecimal) other.value);
+    default:
+      throw new UnsupportedOperationException();
```
Please add error message.

arina-ielchiieva commented on exec/java-exec/src/main/java/org/apache/drill/exec/store/easy/json/parser/RootParser.java:
```diff
@@ -40,13 +44,31 @@ public RootParser(JsonStructureParser structParser) {
+  // report EOF. Use parseRoot() instead.
   @Override
   public void parse(TokenIterator tokenizer) {
     throw new UnsupportedOperationException();
```
Add error message please instead of comment in code.

arina-ielchiieva commented on contrib/storage-http/src/main/java/org/apache/drill/exec/store/http/filter/ConstantHolder.java:
```diff
@@ -375,7 +375,10 @@ public int compareTo(ConstantHolder other) {
-      throw new UnsupportedOperationException();
+      throw new UnsupportedOperationException(
+          String.format(
+              "Unsupported comparison bewteen types %s and %s. Convert values first.",
```
```suggestion
              "Unsupported comparison between types %s and %s. Convert values first.",
```

arina-ielchiieva commented:
@paul-rogers there is one grammatical mistake, other than that PR is good to go.
+1

paul-rogers commented:
@arina-ielchiieva, thanks much for the review! Going to give @cgivre a chance to review it before commit.

cgivre commented:
@arina-ielchiieva, @paul-rogers 
I'm going to do a functional review today so we should be able to commit either this evening or tomorrow.

paul-rogers commented:
Turns out the in-process test server is somewhat limited. Trying out the changes on a "live" site uncovered a few issues. Also, filters were not being reliably pushed down due to limitations of Drill's cost model. New commit addresses these issues, adds tests and cleans up the code a bit.
